Refactor generic thermostat

DB-CL opened the pull request:
## Description

This is a new PR following this one https://github.com/home-assistant/home-assistant/pull/27833 (sorry for taking so much time). I've been instructed to make a PR with breaking changes instead of creating a new component. So here I go.

This thermostat is an amelioration of the current generic_thermostat (better support of preset_mode, hvac_mode and more generally a better compliance with climate 1.0)

## Breaking change
This PR will break the current configuration of generic_thermostat. I tried to keep the way it was but some declarations are not the same. However, this thermostat has at least the same feature than the previous one. Things possible with the previous one should also be possible with this one, you will just have to update your configuration properly.

Old configuration :
```yaml
climate:
  - platform: generic_thermostat
    name: Study
    heater: switch.study_heater
    target_sensor: sensor.study_temperature
    min_temp: 15
    max_temp: 21
    ac_mode: false
    target_temp: 17
    cold_tolerance: 0.3
    hot_tolerance: 0
    min_cycle_duration:
      seconds: 5
    keep_alive:
      minutes: 3
    initial_hvac_mode: "off"
    away_temp: 16
    precision: 0.1
```
New configuration:
```yaml
climate:
  - platform: generic_thermostat
    name: Fake thermostat 4
    sensor: sensor.fake_sensor_1
    initial_hvac_mode: "off"
    initial_preset_mode: "none"
    hysteresis_tolerance_on: 0.5
    hysteresis_tolerance_off: 1
    restore_from_old_state: True
    keep_alive:
      minutes: 3
    min_cycle_duration:
      minutes: 5
    enabled_presets:
      - away
      - eco
      - comfort
    heat:
      entity_id: switch.fake_heater_switch
      min_temp: 15
      max_temp: 24
      initial_target_temp: 19
      away_temp: 12
      eco_shift: -2
      comfort_shift: 2
    cool:
      entity_id: switch.fake_ac_switch
      min_temp: 10
      max_temp: 32
      initial_target_temp: 28
      away_temp: 28
      eco_shift: 4
      comfort_shift: -2
```

## Proposed change
This thermostat support AC mode and HEAT mode in the same component, so you don't have to have two instances to manage the temperature in only one room. It also supports 3 presets modes : Away, Eco and Comfort and solve a few problems we had with the previous one.

It should solve, or at least propose an alternative solution for these issues :
#27350
#27069
#21125
#20154

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:

```yaml
default_config:

frontend:

logger:
  default: info
  logs:
    homeassistant.components.generic_thermostat: debug

input_boolean:
  fake_heater_input:
    name: Fake heater input
    initial: off
  fake_ac_input:
    name: Fake ac input
    initial: off

input_number:
  fake_sensor_temperature_1:
    initial: 20
    min: -20
    max: 35
    step: 0.5

sensor:
  - platform: template
    sensors:
      fake_sensor_1:
        unit_of_measurement: "degrees"
        value_template: "{{ states('input_number.fake_sensor_temperature_1') | float }}"
    
switch:
  - platform: template
    switches:
      fake_heater_switch:
        value_template: "{{ is_state('input_boolean.fake_heater_input', 'on') }}"
        turn_on:
          service: input_boolean.turn_on
          entity_id: input_boolean.fake_heater_input
        turn_off:
          service: input_boolean.turn_off
          entity_id: input_boolean.fake_heater_input
      fake_ac_switch:
        value_template: "{{ is_state('input_boolean.fake_ac_input', 'on') }}"
        turn_on:
          service: input_boolean.turn_on
          entity_id: input_boolean.fake_ac_input
        turn_off:
          service: input_boolean.turn_off
          entity_id: input_boolean.fake_ac_input

climate:
  - platform: generic_thermostat
    name: Fake thermostat 4
    sensor: sensor.fake_sensor_1
    initial_hvac_mode: "off"
    initial_preset_mode: "none"
    hysteresis_tolerance_on: 0.5
    hysteresis_tolerance_off: 1
    restore_from_old_state: True
    keep_alive:
      minutes: 3
    min_cycle_duration:
      minutes: 5
    sensor_stale_duration:
      minutes: 120
    heat:
      entity_id: switch.fake_heater_switch
      min_temp: 15
      max_temp: 24
      initial_target_temp: 19
      away_temp: 12
      eco_shift: -2
      comfort_shift: 2
    cool:
      entity_id: switch.fake_ac_switch
      min_temp: 10
      max_temp: 32
      initial_target_temp: 28
      away_temp: 28
      eco_shift: 4
      comfort_shift: -2
```

## Additional information
- This PR fixes or closes issue: 
#27350
#27069
#21125
#20154
- This PR is related to this PR: https://github.com/home-assistant/home-assistant/pull/27833
- Link to documentation pull request: documentation will come when the PR will be stabilized
- I have a 100% test coverage by running : `pytest tests/components/generic_thermostat/test_generic_thermostat.py --cov=homeassistant/components/generic_thermostat/ --cov-report term-missing`

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository] : https://github.com/home-assistant/home-assistant.io/pull/13384

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

DB-CL committed:
Squashing commits to get the rebase with dev working

Misiu commented:
@DB-CL so the name of the new component is `virtual_thermostat` or will it still be `generic_thermostat` (sample config is using virtual, but in code, it looks like generic)?
How can we set ECO and COMPORT temperatures? I see that AWAY is configurable directly: https://github.com/home-assistant/home-assistant/pull/31799/files?file-filters%5B%5D=.py&hide-deleted-files=true#diff-2bc6470dbfbdc14a4cfacc4ad731a410R72 but ECO and COMFORT use shift. Is this correct?

DB-CL commented:
@Misiu Thank you for pointing that out, it was a copy/paste mistake from my last PR's description. And yes, the away preset is a fixed temperature whereas the comfort and eco are shifts from the setpoint. The setpoint is for the "None" preset and is set in the initial_target_temp.

@frenck Should I spend time on the PR documentation right now ? Because I have the feeling that this PR will be subject to modifications and I don't have the time to update the documentation each time the PR will be updated.

Misiu commented:
@DB-CL thanks for the clarification.
I must admit this is a bit confusing. Two values are passed as a value and two as an offset.
```
initial_target_temp: 19
away_temp: 12
eco_shift: -2
comfort_shift: 2
````

If I set values in my config as above, then change the target temp with UI to 22 what will be the target value of ECO? Will it be 17 (19-2) or 20 (22-2)? Also what happens after the restart of HA? What will be the target temperature (assuming I selected ECO preset before restart)?
If I select ECO or COMFORT with the UI then the slider will be disabled? So I won't be able to change the target temperature (until I change preset to NONE or AWAY).
I only used the old `generic_thermostat`, never had one with presets so I don't know what to expect.
I've tried the demo (https://demo.home-assistant.io/#/lovelace/0) and changing the preset doesn't affect the target temperature.

DB-CL commented:
> I must admit this is a bit confusing. Two values are passed as a value and two as an offset.

I understand, but I didn't make it up. This is how a thermostat works in most cases, at least in my country. The away temp is often a low limit, that's why it's a fixed value. I tried to differentiate the configuration values by putting a _shift and a _temp, if you think of something better I can make some changes.

> If I set values in my config as above, then change the target temp with UI to 22 what will be the target value of ECO

This question is not complete. You didn't specified the preset mode you are currently running. If you are in ECO mode, then the temperature displayed in the UI will be the ECO one, so if you put 22 in the UI it will be 22 in ECO. If you do something like that, then when you will switch back to "no preset" the target temp will be 24 and if you switch again to COMFORT the target temp will be 26.

For your information, in case you find this behavior strange, I didn't choose to display the target temp in the UI (I didn't made the UI) and I didn't choose the definition of the target temp, which is the temperature the thermostat is currently trying to reach (here :  https://developers.home-assistant.io/docs/en/entity_climate.html ).

> Also what happens after the restart of HA?

If you set the `restore_from_old_state` to `true` then the state will be fully restored (preset, target, mode, ..) and if you set it to `false` then the initial values set in the configuration will be used

> If I select ECO or COMFORT with the UI then the slider will be disabled?

No. Again, this PR is not about the UI. The UI will be the same as before. If you select ECO, it will change the target temp in the slider by the eco shift, and also change the boundaries, same with COMFORT. If you manipulate the slider when you are in ECO, COMFORT or NONE preset, then the target temperature will be changed. If you select AWAY, I've reduced the range of temperature to 0 so you won't be able to move the slider, this is a choice I made because changing the target temp with absolutely zero effect would be very strange and counter intuitive.

You can play with the UI very easily by pulling my PR and launching hass ( `hass -c config` ) with the configuration file I gave in the PR. Hass will be accessible on 127.0.0.1:8123

I don't think the presets are meant to be set manually in climate 1.0. They are meant to be manipulated via automations. The UI does not allow to change the PRESET values ... just the target value. I think that would be great if we could change the preset value of COMFORT, ECO, etc.. directly in the UI.

Another way to go would be to have 3 totally different target temperatures per mode and the slider could allow us to change one or the other, I think that would be possible but I must dig into the code because the state restoration could be a problem since the target temp is supposed to be unique. What others have to say about that ?

dgomes commented on .gitignore:
```diff
@@ -131,3 +131,5 @@ tmp_cache
 # python-language-server / Rope
 .ropeproject
+
+home-assistant-polymer
```
Probably should be out of this PR

DB-CL replied:
I'm removing it

dgomes commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -43,84 +45,190 @@
+    enabled_hvac_modes = []
+
+    # Do all the checks I cannot do with voluptuous
+    if initial_preset_mode != "none" and initial_preset_mode not in enabled_presets:
```
You should follow homeassistant.helpers.config_validation as an example to do this check

DB-CL replied:
Thanks for this information, is this the code located in this file ? https://github.com/home-assistant/home-assistant/blob/master/homeassistant/helpers/config_validation.py

Because I've read it now, and I still don't know what I'm supposed to do. Can you point me in the right direction ?

dgomes replied:
I didn't quite get your comment on L160 

Can't you extent the config_validation with new functions to do these checks through voluptuous ?

DB-CL replied:
Unfortunately I have no idea how I can do such thing. If you know a component that do something like that, please share its name with me so I can have an example of what I'm supposed to do.

dgomes replied:
Just picking up a reference to the first extension in config_validation.py (https://github.com/home-assistant/home-assistant/blob/75b508e8724cb82712227082eee9abab570d2eb2/homeassistant/helpers/config_validation.py#L94)

In azure_service_bus component this extra validation is used: 

https://github.com/home-assistant/home-assistant/blob/75b508e8724cb82712227082eee9abab570d2eb2/homeassistant/components/azure_service_bus/notify.py#L32

DB-CL replied:
Thanks, I did an attempt to implement the first check in voluptuous as a custom validator. It's working. Tell me if it is what you are expecting, and I will do it for the next checks.

DB-CL replied:
I did all the checks with voluptuous

dgomes commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -132,336 +240,530 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+            return
+        data = {ATTR_ENTITY_ID: self._heater_entity_id}
+        _LOGGER.debug("Order OFF sent")
+        _LOGGER.debug("Order OFF sent to heater device %s", self._heater_entity_id)
```
Duplicate debug ?

DB-CL replied:
Yes thank you for pointing this out.

dgomes reviewed (commented):
Maybe debug lines can be trimmed down a bit after development

dgomes commented on homeassistant/components/generic_thermostat/manifest.json:
```diff
@@ -1,8 +1,11 @@
+        "sensor",
+        "switch"
+    ],
+    "codeowners": []
```
Care to add yourself :) ?

pvizeli replied:
Yeah, need to be add @DB-CL

pvizeli replied:
This is recommended

DB-CL replied:
Done

mikenabhan commented:
I see this got marked as stale, it looks like it is missing the documentation pull.  I saw this in the original pull request for docs, I could take a crack at updating if it would help move this forward...

https://github.com/home-assistant/home-assistant.io/pull/10831

DB-CL commented:
Hi @mikenabhan 

It's not really missing the doc, as you said the doc is there, I just have to move it and every body knows that. I don't do it because I don't have much time and I don't want to waste it.

It's missing a proper discussion about this PR... 

I'm introducing breaking changes, and except @Misiu, nobody seems to really care about the changes. There is no discussion at all although the things @Misiu said are true. And I think there is even more to discuss about.

I don't want this PR to be merged if nobody really understand the changes, otherwise it could be a disaster.

You can try to get people involved in this, I will provide answers and try to make changes as best as I can.

Misiu commented:
@DB-CL you did a massive work.
Maybe try asking on discord for a review.
Hopefully, this will bring more attention 🤞 and this can move forward.

dgomes commented:
Having documentation and fixing the conflicts would really help make the case for a second reviewer to come by and review

edofullin commented:
I'm using your virtual thermostat in my home assistant as a custom component and except some minor bugs here and there has been working without problems all winter.

I hope this can get along, thanks for your work

pvizeli reviewed (commented):
```
 enabled_presets:
      - away
      - eco
      - comfort
```

ditch this options

pvizeli reviewed (approved):
Address my 2 comments and we can merge it

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+            hvac_mode = self._hvac_mode
+        elif hvac_mode not in self.hvac_modes:
+            _LOGGER.warning(
+                "Try to update temperature to %s for mode %s but this mode is not enabled. Skipping.",
```
We don't end logging messages with period.
```suggestion
                "Ignoring update temperature to %s for mode %s as this mode is not enabled,
```

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+            return
+
+        await self._async_operate()
+        await self.async_update_ha_state()
```
```suggestion
        self.async_write_ha_state()
```

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
-        await self._async_control_heating()
+        self._async_update_current_temp(new_state)
+        await self._async_operate(sensor_changed=True)
         await self.async_update_ha_state()
```
```suggestion
        self.async_write_ha_state()
```

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+            current_temp = self._current_temperature
+
+            _LOGGER.debug(
+                "Operate - target_temp_min %s, target_temp_max %s, current_temp %s, target_temp %s, shift %s, keepalive %s",
```
Please break long strings around 88 characters.

DB-CL replied:
I've break the line

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+            self._preset_mode = preset_mode
+            _LOGGER.debug("Set preset mode to %s", preset_mode)
+            await self._async_operate()
+            await self.async_update_ha_state()
```
```suggestion
            self.async_write_ha_state()
```

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        if preset_mode in self.preset_modes or preset_mode == PRESET_NONE:
```
Invert this check and log error and return if true. Then we can outdent below.

DB-CL replied:
Done

MartinHjelmare reviewed (changes requested):
Please name the test module `test_climate.py`.

MartinHjelmare commented:
Please rebase on latest dev branch and solve the merge conflicts.

DB-CL commented:
Hi, thank you for your comments, I'll look into it soon.

pvizeli commented:
I think we should just set eco/away/conform without shift and hold it simple. If it is set per config, we show it up in the list otherwise not.

Misiu commented:
@DB-CL I'd like to ask one more thing about your PR, to be sure it fixes one big issue with the current implementation. 
There are many threads in forum about generic thermostat not turning off when the target temperature sensor if offline or report bad reading. For example this one: https://community.home-assistant.io/t/generic-thermostat-sensor-unavailable-should-go-off/80909 or this: https://community.home-assistant.io/t/generic-thermostat-failovers-with-tasmota/73087

How will your implementation work? I think it should turn the thermostat off. This would be a great feature because now you can damage your home by not turning off the heating.

ncortines commented:
I know I'm a bit late to the party, but I have a question: Does the new implementation cater for the situation where the sensor becomes unavailable while on HEAT mode? Currently ```generic_thermostat``` maintains HEAT status when the sensor becomes unavailable. It'd be nice to change this behaviour or at least provide some additional configuration option to handle it. Thanks :)

DB-CL commented:
Hi all, 

Here is a list of what I need to do

- [ ] Work on "enabled_presets" option ( @pvizeli )
- [ ] Rename the test module ( @MartinHjelmare )
- [ ] Check the comments of @MartinHjelmare about the code ( many thanks for the deep review )
- [ ] Discuss about the eco/away/comfort shift ( @pvizeli )
- [ ] Do a PR for the documentation ( @pvizeli )
- [ ] Check the behaviour when the sensor goes offline ( @Misiu, @ncortines )
- [ ] Catch up with reality ( @MartinHjelmare )

I'll give it a try this evening

vindaalex commented:
One question that pops up for me is as well the safety aspect, like the switch and temp reading. For instance a safety that puts it in idle mode when the temperature reading is far too high (this may indicate an offline switch/damaged valve) or the switch is unavailable. Maybe the the option to put in a secondary temp reading as back up in case the primary temp sensor crashes and keeps giving a constant value.

Shulyaka commented:
You can implement a secondary sensor with `template_sensor`, no need to overcomplicate this integration.

DB-CL commented:
To start, I'm sorry but I need clarifications about what I should do with comfort/eco/away.

I got the point about the `enabled_presets` options and I'll drop this to make an auto detection. But I don't get the point about the "shift" option. 

> I think we should just set eco/away/conform without shift and hold it simple. If it is set per config, we show it up in the list otherwise not.

I'm confused with the second sentence. Show what in what list ? Are you talking about the `enabled_presets` that you want me to remove ? In that case what is the matter with the shift ?

I still think there should be a discussion that @Misiu started about the shift option ( shift vs plain ) because I've digged into the code, and as far as I understand, you cannot store plain temperature because the state restoration won't be able to restore it. It can restore only the target temperature. So if we want to be able to change the temperature via the GUI when the thermostat is on a preset mode, the only way I found is to use shifts. I know it's tricky and I'm not very clear because of the language barrier, that's why I would like to discuss it with someone with knowledge into that component and with patience about that problematic (even outside github if needed).

Thanks for pointing me in the right direction

DB-CL commented:
> You can implement a secondary sensor with `template_sensor`, no need to overcomplicate this integration.

Good point, but that's somehow hard to get for newcomers. Should we talk about this in the documentation for example ?

Shulyaka commented:
> Good point, but that's somehow hard to get for newcomers. Should we talk about this in the documentation for example ?

Probably. But I believe this is not the first priority.

vindaalex commented:
For the temp sensor backup the template_sensor seems to be a good option. What would happen when the switch state changes to unavailable?

Shulyaka commented:
> For the temp sensor backup the template_sensor seems to be a good option. What would happen when the switch state changes to unavailable?

I think that does not matter, if the switch is unavailable we cannot control anything anyway.

DB-CL commented:
Hi, 

I'm going to need help :( I'm stuck in a git shit storm. It's too bad, I've just spent a couple of hours on this instead of working on the PR :'(

Here is the story : I changed my setup and thought I could just pull my PR from github and continue working on.

To pull my PR I do the following : 

```bash
$ git clone https://github.com/home-assistant/core.git home-assistant
$ cd home-assistant
$ git fetch origin pull/31799/head:new_generic_thermostat
$ git checkout new_generic_thermostat
```
At this point I have my PR clean, and I'm on the right branch. Everything is fine.

To rebase I'm following this guide : https://developers.home-assistant.io/docs/development_catching_up/ and so far, it has been working.

So I'm running : 
```bash
$ git remote add upstream https://github.com/home-assistant/core.git
$ git fetch upstream dev
$ git rebase upstream/dev
```

And the problems are starting. I'm having conflicts with code totally unrelated to my last changes. As far as I understand, it's trying to apply old commits of mine to the current code base, but it looks like the conflicts found are between my own commits (the "them" looks like it's "me" in the three ways merge). It's a real nightmare and I tried many times to start from scratch and resolve the conflicts, but in the end I'm always stuck with conflicts errors :(

I feel like I should squash my commits and try to rebase with that. Is that the right way to do ? Is that clean ? I'm not even sure it will help.

DB-CL committed:
Changes requested & sensor failure management

DB-CL commented:
So, 

- I've removed the `enabled_presets` and put an automatic detection in place (with all the checks preserved). The code is less "clean" imo but the configuration is easier
- I've added something to manage a sensor failure : `sensor_stale_duration`. The component will check if the state of the sensor has been updated during this period. If not, it will go into emergency mode (turning OFF the device of the current active mode). If the sensor has a "None" state, it will also trigger this emergency mode (this one is non configurable, so it will not depend on the sensor_stale_duration variable). The only way a thermostat goes in emergency mode, it's because a sensor failure (None state or stalled sensor) so the only way for it to goes out this emergency mode is to get a new clean value from the sensor.

There is a little problem with this new check : I don't know what happen when a sensor get the same value from the physical device in the real world. In the unit test, I had to force the state value (`force_update=True`) to make it work, but in the real world I don't know what really happen. I personnally think that the same value is a new information so the state should be changed but I don't know it for sure. Anyway, I think it's a minor case (the sensor keep sending the same value for a long period ..) and if it bother people, they should'nt use this `sensor_state_duration` feature. It will be documented ofc.

DB-CL committed:
Codespell & codeowners

DB-CL committed:
Linting

DB-CL committed:
Linting

MartinHjelmare commented:
I don't think we should add the emergency mode in this PR. That needs more discussion first.

ncortines commented:
> You can implement a secondary sensor with `template_sensor`, no need to overcomplicate this integration.

Sorry, newcomer here :) Could you please elaborate a bit? I'd like to know what is the recommended way to overcome the situation in which the sensor becomes unavailable while thermostat is on heat mode :) Thanks

DB-CL commented:
> Sorry, newcomer here :) Could you please elaborate a bit? I'd like to know what is the recommended way to overcome the situation in which the sensor becomes unavailable while thermostat is on heat mode :) Thanks

This is not the case covered by the suggestion. It was about a sensor having very high values. This PR is covering the case when a sensor goes offline ( no new values ) or when a state is wrong ( "None" state ). This is the `emergency_mode`.

> I don't think we should add the emergency mode in this PR. That needs more discussion first.

That couldn't be discussed right now ? I can trim the code off but you'll have to reimplement this soon, this is a feature I've been asked for many times.

DB-CL commented:
I've added the documentation.

vindaalex commented:
I think, but just a user not developer, that safety should be a priority for these kind of automations. The use of template_sensor is a good option to make the temp input more robust. But (I think) eventually the thermostat should switch off/idle/emergency mode when it notices  something 'strange'. Or could this be achieved with an external automation?

MartinHjelmare commented:
We should keep this first PR as simple as possible and only address the retain and restore features that were missing and were the reason for the issues that we're trying to address with this PR.

Anything else should be one or many future PRs.

fjufju commented:
I disagree with @MartinHjelmare and agree with @vindaalex.
Security should be first.
Now implementation can damage home. Must be addressed.

Shulyaka commented:
> I disagree with @MartinHjelmare and agree with @vindaalex.
> Security should be first.
> Now implementation can damage home. Must be addressed.

But this PR does not decrease security. Security is important, but doesn't have to be handled in a bundle with other features.

Shulyaka commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
 
 
-class GenericThermostat(ClimateEntity, RestoreEntity):
+class GenericThermostat(ClimateDevice, RestoreEntity):
```
```suggestion
class GenericThermostat(ClimateEntity, RestoreEntity):
```

Shulyaka commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -1,11 +1,13 @@
 import voluptuous as vol
 
-from homeassistant.components.climate import PLATFORM_SCHEMA, ClimateEntity
+from homeassistant.components.climate import PLATFORM_SCHEMA, ClimateDevice
```
```suggestion
from homeassistant.components.climate import PLATFORM_SCHEMA, ClimateEntity
```

Shulyaka commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
-            ):
-                self._async_update_temp(sensor_state)
+            sensor_state = self.hass.states.get(self._sensor_entity_id)
+            if sensor_state and sensor_state.state != STATE_UNKNOWN:
```
```suggestion
            if sensor_state and sensor_state.state not in (
                STATE_UNAVAILABLE,
                STATE_UNKNOWN,
            ):
```

Shulyaka commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
         self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, _async_startup)
 
-        # Check If we have an old state
+        _async_startup(None)  # init the sensor
```
Shouldn't we call it after we restore the state? The function calls `self._async_update_current_temp` which might need the correct state.

Shulyaka commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
+
+        if sensor_state.last_updated < now - self._sensor_stale_duration:
+            _LOGGER.debug("Time is %s, last changed is %s, stale duration is %s")
+            _LOGGER.debug("Sensor is stalled, call the emergency stop")
```
```suggestion
            _LOGGER.warning("Sensor is stalled, call the emergency stop")
```

blademckain commented:
congratulations great job!
expect to see it within the main release

my 2 cents:
add the possibility to control other thermostats (master + slave), inserting in the configuration
```yaml
mode: master or standalone
slave:
    - climate.clima1
    - climate.clima2
    - climate.clima3
```

in master mode any change in the gui is propagated to the slaves. 
you can grab code from [here](https://github.com/daenny/climate_group)

MartinHjelmare commented:
@DB-CL what's the status? Are you planning to address the review comments? Otherwise we should close this PR. We're trying to decrease our open PR buffer. Thanks!

Shulyaka commented:
> add the possibility to control other thermostats (master + slave), inserting in the configuration

I think that should be a different integration.

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
+            vol.Optional(CONF_MIN_CYCLE_DURATION): vol.All(
+                cv.time_period, cv.positive_timedelta
+            ),
+            vol.Optional(CONF_STALE_DURATION): vol.All(
```
Please remove this option.

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
+        self._tolerance_off = tolerance_off
+        self._unit = unit
+        self._keep_alive = keep_alive
+        self._sensor_stale_duration = sensor_stale_duration
+        self._emergency_stop = False
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
 
-        if self._keep_alive:
+        if self._sensor_stale_duration:
             async_track_time_interval(
-                self.hass, self._async_control_heating, self._keep_alive
+                self.hass,
+                self._async_check_sensor_not_responding,
+                self._sensor_stale_duration,
             )
```
Please remove this.

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
+        await self._async_operate(sensor_changed=True)
         self.async_write_ha_state()
 
+    async def _async_check_sensor_not_responding(self, now=None):
```
Please remove this.

MartinHjelmare commented on homeassistant/components/generic_thermostat/climate.py:
```diff
@@ -44,428 +46,813 @@
+        _LOGGER.debug("Order OFF sent to AC device %s", self._ac_entity_id)
+        await self.hass.services.async_call(HA_DOMAIN, SERVICE_TURN_OFF, data)
+
+    async def _activate_emergency_stop(self):
```
Please remove this.

tibortiz commented:
> > add the possibility to control other thermostats (master + slave), inserting in the configuration
> 
> I think that should be a different integration.

Is it a way to list all the possible future improvements for the generic thermostat ?
On my side I will be quite happy to have the possibility to link the generic_thermostat with a person entity, person group or any other sensor (with 'home' and 'not_home' state) to trigger the away/home preset of the generic thermostat.
I'm using workaround with multiple automation to do it, it's a bit dirty but it works.

edofullin commented:
I'm really interested in this PR. 

If @DB-CL does not have the time or will to continue working on this, I can address the requested changes to take this PR forward.

Please let me know

DB-CL commented:
Hi !

Thank you very much for your comments and reviewing. As I already mentioned, I have very limited amount of time for my personal projects that's why I'm not often here. I also let the time to other people to write because when I'm too quick, I end up removing perfectly working and useful pieces of code ;)

I will have time to work on this PR at the end of july I think, but unfortunately not before. Until then, feel free to continue the work. I'm not sure if I can open this PR to multiple people though :-/

edofullin commented:
Another thing: I wouldn't force an off state to the heat/cool switch if the thermostat is turned off. 
For example if the thermostat is turned off and a member of my family tries to turn on the A/C manually with the remote, after a few seconds the thermostat shuts down the A/C anyway.

If the thermostat is on heat/cool modes that's fine but if it's turned off I understand it as a "thermostat not operating" not as a "force off all devices"




> congratulations great job!
> expect to see it within the main release
> 
> my 2 cents:
> add the possibility to control other thermostats (master + slave), inserting in the configuration
> 
> ```yaml
> mode: master or standalone
> slave:
>     - climate.clima1
>     - climate.clima2
>     - climate.clima3
> ```
> 
> in master mode any change in the gui is propagated to the slaves.
> you can grab code from [here](https://github.com/daenny/climate_group)

I think that could be a different integration, and can be made platform-independent

fjufju commented:
Why close? This is important. Author wrote he will finish. Please open again. Thank you :)

MartinHjelmare commented:
It's stale. We're trying to decrease our open PR buffer. The author can ask to reopen when there's time to continue.

fjufju commented:
Ok. Sorry for trouble @MartinHjelmare.

@DB-CL please open when have time

shlomki commented:
I too would love to have these new features. Crossing fingers that @DB-CL would have the time to finish it :)

blademckain commented:
I too would love to have these new features.
+1

benlbrm commented:
@DB-CL Please come back ! we definitively needs this !
;-)
